# Mapping: Summary Statistics

## Genomic Summary

### Number of Genomic Hits
```{r pkgs-part2, eval=FALSE}
library(mapstatHelpers)
```

Import the data on assigned reads.
This can take a while as quite a few files need to be read
```{r big-import}
gen_mapstats <- 
  read_multiple_mapstats("../kma_EFFORT/data/final_mapstats/livestockAMR_Genomic_mapstats/") %>% 
  filter(sample_id %in% include_samples) %>% 
  mutate(id = str_extract(refSequence, "[:graph:]+"))

gen_refdata <- read_refdata("../kma_EFFORT/data/2019-09-09-combined_genomic_20190404.refdata") %>% 
  mutate(custom_group = case_when(db_name == "IMGVR_20190128" ~ "Phages",
                                  db_name == "parasite_cge_20190404" ~ "Parasites",
                                  db_name == "fungi_20190213" ~ "Fungi",
                                  db_name == "human_20181105" ~ "Human",
                                  db_name == "protozoa_20190213" ~ "Protozoa",
                                  db_name == "kvit_20190205" ~ "Viruses",
                                  superkingdom_name == "NULL" &
                                    db_name %in% c("MetaHitAssembly_20140701",
                                                   "HumanMicrobiome_20140702") ~ "Other Bacteria/Archaea",
                                  superkingdom_name == "Eukaryota" ~ "Other Eukaryotes",
                                  T ~ superkingdom_name))

bac_long <- gen_mapstats %>%
  left_join(select(gen_refdata, id, ends_with("_name"))) %>%
  filter(superkingdom_name == "Bacteria") %>%
  arrange(sample_id, id)
```
Let us take a look at how many of the reads were mapped to the known genomes
```{r}
mapped_to_gen <- gen_mapstats %>% 
  group_by(sample_id) %>% 
  summarise(total_mapped = sum(fragmentCount)) %>% 
  left_join(livestock_sample_metadata) %>% 
  mutate(mapped_pct = 100 * total_mapped / total_fragments)
```

```{r}
mapped_to_gen %>% 
  ggplot(aes(x = country, y = mapped_pct, fill = species)) +
  geom_boxplot(outlier.shape = 21) +
  facet_grid(cols = vars(species), scales = "free_x", space = "free_x") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Country", y = "% of fragments mapped to Genomic") +
  ggtitle("Proportion of Mapped Fragments (against Genomic DB)") +
  guides(fill = F) +
  theme_light() +
  theme(panel.grid.major.x = element_blank())
```

### Genomic Sub-DB Composition
```{r}
db_mapping_count <- gen_mapstats %>% 
  left_join(select(gen_refdata, id, db_name)) %>% 
  group_by(sample_id, db_name) %>% 
  summarise(fragments_per_db = sum(fragmentCount)) %>% 
  ungroup() %>% 
  complete(db_name, sample_id, fill = list(fragments_per_db = 0))

unmapped_df <- db_mapping_count %>% 
  group_by(sample_id) %>% 
  summarise(total_mapped = sum(fragments_per_db)) %>% 
  left_join(livestock_sample_metadata) %>% 
  ungroup() %>% 
  mutate(unmapped_fragments = total_fragments - total_mapped,
         db_name = "Unmapped") %>% 
  select(sample_id, fragments_per_db = unmapped_fragments, db_name)

db_proportions_summary <- bind_rows(db_mapping_count, unmapped_df) %>% 
  left_join(livestock_sample_metadata) %>% 
  mutate(db_pct = 100 * fragments_per_db / total_fragments)

interesting_dbs <- c("bacteria_20190205", "HumanMicrobiome_20140702", 
                     "MetaHitAssembly_20140701", "bacteria_draft_20190205", 
                     "plasmid_20190205")
```

```{r}
db_proportions_summary %>% 
  filter(db_name %in% interesting_dbs) %>% 
  mutate(db_name = str_replace(db_name, "_\\d+", "")) %>% 
  ggplot(aes(x = explicit_country, y = db_pct, fill = species)) +
  geom_boxplot(outlier.shape = 21) +
  facet_grid(cols = vars(species), rows = vars(db_name), scales = "free", space = "free_x") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Country", y = "% of fragments mapped") +
  ggtitle("Proportion of Mapped Fragments (against Genomic DB)") +
  guides(fill = F) +
  theme_light() +
  theme(panel.grid.major.x = element_blank())
```

```{r, results='markup'}
db_proportions_summary %>% 
  filter(!db_name %in% interesting_dbs, db_name != "Unmapped",
         db_pct > 1) %>% 
  mutate(m_fragments = fragments_per_db * 1e-06) %>% 
  mutate_if(is.numeric, round, digits = 2) %>% 
  select(sample_id, db = db_name, species, explicit_country, m_fragments, db_pct) %>% 
  arrange(desc(db_pct)) %>% 
  DT::datatable(., rownames = F)
```


## ResFinder Summary

```{r}
res_mapstats <- read_multiple_mapstats("../kma_EFFORT/data/final_mapstats/ResFinder_20190213/")

res_long <- res_mapstats %>% 
  filter(sample_id %in% include_samples) %>% 
  mutate(gene_id = str_extract(refSequence, "[:graph:]+")) %>% 
  select(-refSequence) %>% 
  left_join(resfinder_metadata) %>% 
  select(sample_id, gene_id, amr_class, everything()) %>% 
  mutate(frags_per_bp = fragmentCount / gene_length_bp) %>% 
  mutate(pct_covered = 100 * refConsensusSum / gene_length_bp) %>% 
  filter(pct_covered >= 20)
```

### Number of ResFinder Hits

```{r}
resfinder_mapped_prop <- res_long %>% 
  group_by(sample_id) %>% 
  summarise(res_mapped_fragments = sum(fragmentCount)) %>% 
  ungroup() %>% 
  bind_rows(data.frame(sample_id = setdiff(include_samples, res_long$sample_id),
                       res_mapped_fragments = 0,
                       stringsAsFactors = F)) %>%
  left_join(livestock_sample_metadata) %>% 
  mutate(resf_pct = 100 * res_mapped_fragments / total_fragments)
```

```{r}
p_respct <- resfinder_mapped_prop %>%
  ggplot(aes(x = explicit_country, y = resf_pct, fill = species)) +
  geom_boxplot(outlier.shape = 21) +
  facet_grid(cols = vars(species), scales = "free_x", space = "free_x") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Country", y = "% ResFinder hits") +
  # ggtitle("ResFinder mapping proportion") +
  guides(fill = F) +
  theme_light(base_size = 9) +
  theme(panel.grid.major.x = element_blank())

p_respct
```

### Comparisons to First Results

The pigs and the broilers were also analyzed with a different pipeline back in the Munk (2018) study. Let us compare the current results with those.

```{r}
bact_hits <- db_mapping_count %>% 
  filter(db_name %in% c("plasmid_20190205", 
                        "bacteria_draft_20190205", 
                        "HumanMicrobiome_20140702", 
                        "MetaHitAssembly_20140701", 
                        "bacteria_20190205", 
                        "archaea_20190213")) %>% 
  group_by(sample_id) %>% 
  summarise(bact_fragments = sum(fragments_per_db))
```

```{r total-abundance-plot}
res_long %>% 
  left_join(bact_hits) %>% 
  mutate(fully_normalised = frags_per_bp / bact_fragments) %>%
  group_by(sample_id) %>% 
  mutate(total_abundance = sum(fully_normalised)) %>% 
  distinct(sample_id, total_abundance) %>% 
  left_join(livestock_sample_metadata) %>% 
  filter(species %in% c("Pig", "Broiler")) %>%
  ggplot(aes(x = explicit_country, y = total_abundance, fill = country)) +
  geom_boxplot(outlier.shape = 21) +
  facet_wrap(. ~ species, scales = "free", nrow = 1) +
  scale_fill_manual(values = plot_colours$country) +
  labs(x = "Country", y = "AMR load - exp(ALR)") +
  ggtitle("Total AMR Abundance") +
  guides(fill = F) +
  theme_light() +
  theme(panel.grid.major.x = element_blank())
```
Let us calculate the additive log ratio (ALR)

```{r total-alr-plot}
p_totalALRbox <- res_long %>% 
  left_join(bact_hits) %>% 
  mutate(fully_normalised = frags_per_bp / bact_fragments) %>%
  group_by(sample_id) %>% 
  mutate(total_abundance = log2(sum(fully_normalised))) %>% 
  distinct(sample_id, total_abundance) %>% 
  left_join(livestock_sample_metadata) %>%
  ggplot(aes(x = explicit_country, y = total_abundance, fill = species)) +
  geom_boxplot(outlier.shape = 21) +
  facet_grid(cols = vars(species), scales = "free_x", space = "free_x") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Country", y = "ALR (log2)") +
  guides(fill = F) +
  theme_classic(base_size = 9) +
  theme(panel.border = element_rect(linetype = "solid", fill = NA),
        strip.background = element_blank(),
        axis.line = element_blank())

p_totalALRbox
```

```{r munk2018-fig1b}
p_stackedbar <- res_long %>% 
  group_by(sample_id, munk2018_amr_group) %>% 
  summarise(sum_fragments = sum(fragmentCount)) %>% 
  left_join(livestock_sample_metadata) %>%
  group_by(explicit_country, species) %>% 
  mutate(n_group = dense_rank(sample_id)) %>% 
  ggplot(., aes(x = n_group, y = sum_fragments, fill = munk2018_amr_group)) +
  geom_bar(stat = "identity", position = "fill") +
  scale_fill_manual(values = plot_colours$new_amr_group, guide = guide_legend(title = NULL)) +
  labs(y = "AMR composition") +
  facet_wrap(facets = vars(species, explicit_country), scales = "free_x", ncol = 9,
             labeller = labeller(.multi_line = F)) +
  theme_classic(base_size = 9) +
  theme(panel.border = element_rect(linetype = "solid", fill = NA),
        strip.background = element_blank(),
        axis.line = element_blank(),
        axis.text = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks = element_blank(),
        legend.key.size = unit(0.3, "cm"),
        panel.spacing.x = unit(0.05, "cm"),
        panel.spacing.y = unit(0.1, "cm"),
        legend.box.spacing = unit(0.1, "cm"),
        legend.margin = margin(0.1, 0.1, 0.1, 0.1, "cm"))

p_stackedbar
```

### Alpha Diversity (ResFinder)

```{r}
quick_chao1 <- function(vector_in){
  f1 <- sum(vector_in == 1)
  f2 <- sum(vector_in == 2)
  sum(vector_in > 0) + ((f1 * (f1 - 1)) / (2 * (f2 + 1)))
}

quick_shannon <- function(vector_in){
  vector_use <- vector_in[which(vector_in > 0)]
  p <- vector_use / sum(vector_use)
  -1 * sum(p * log(p))
}

res_alphadiv <- res_long %>% 
  group_by(sample_id) %>% 
  mutate(res_chao1 = quick_chao1(fragmentCount),
         res_shannon = quick_shannon(fragmentCount)) %>% 
  ungroup() %>% 
  select(sample_id, res_chao1, res_shannon) %>% 
  distinct() %>% 
  bind_rows(data.frame(sample_id = setdiff(include_samples, res_long$sample_id),
                       res_chao1 = 0,
                       stringsAsFactors = F))
```

```{r}
p_chao <- res_alphadiv %>% 
  left_join(livestock_sample_metadata) %>% 
  #mutate(species = factor(species, levels = c("Chicken", "Pig", "Turkey", "Veal Calf", "Fish"))) %>% 
  #mutate(species = factor(species, levels = c("Broiler", "Pig", "Turkey", "Veal Calf", "Fish"))) %>% 
  ggplot(aes(x = explicit_country, y = res_chao1, fill = species)) +
  geom_boxplot(outlier.shape = 21) +
  facet_grid(cols = vars(species), scales = "free_x", space = "free_x") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Country", y = "AMR richness (Chao1)") +
  guides(fill = F) +
  theme_classic(base_size = 9) +
  theme(panel.border = element_rect(linetype = "solid", fill = NA),
        strip.background = element_blank(),
        axis.line = element_blank())
# theme_light(base_size = 10) +
# theme(panel.grid.major.x = element_blank(),
#       panel.grid.minor.y = element_blank())

p_chao
```

```{r}
p_evenness = res_alphadiv %>% 
  filter(! is.na(res_shannon)) %>% 
  left_join(livestock_sample_metadata) %>% 
  #mutate(species = factor(species, levels = c("Chicken", "Pig", "Turkey", "Veal Calf", "Fish"))) %>% 
  #mutate(species = factor(species, levels = c("Broiler", "Pig", "Turkey", "Veal Calf", "Fish"))) %>% 
  ggplot(aes(x = explicit_country, y = res_shannon, fill = species)) +
  geom_boxplot(outlier.shape = 21) +
  facet_grid(cols = vars(species), scales = "free_x", space = "free_x") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Country", y = "Shannon index") +
  guides(fill = F) +
  theme_light(base_size = 9) +
  theme(panel.grid.major.x = element_blank())

#ggsave("supp_shannon_evenness.pdf",
#       width = 17.7, height = 6, units = "cm")
p_evenness
```

### Alpha Diversity (Bacterial species)


```{r}
bac_alphadiv <- bac_long %>% 
  select(sample_id, species_name, fragmentCount) %>%
  group_by(sample_id, species_name) %>% 
  summarise(fragmentCount = sum(fragmentCount)) %>%
  mutate(bac_chao1 = quick_chao1(fragmentCount),
         bac_shannon = quick_shannon(fragmentCount)) %>% 
  ungroup() %>% 
  select(sample_id, bac_chao1, bac_shannon) %>% 
  distinct() %>% 
  bind_rows(data.frame(sample_id = setdiff(include_samples, res_long$sample_id),
                       bac_chao1 = 0,
                       stringsAsFactors = F))

#ggsave("supp_shannon_evenness.pdf",
#       width = 17.7, height = 6, units = "cm")
bac_alphadiv
```
### Rarefyed version
```{r}
genus_long %>% 
    pivot_wider(names_from = genus_name, 
                values_from = genus_frags, 
                values_fill = 0) %>%
    data.frame(., row.names = 1) %>% 
    select(-1) %>% 
    vegan::rrarefy(sample = 2000) %>% 
    as_tibble(rownames = "sample_id") %>%
    pivot_longer(cols = -1, names_to = "genus_name", values_to = "genus_frags") %>%
    filter(genus_frags > 0) %>%
    group_by(sample_id) %>%
    mutate(bac_chao1 = quick_chao1(genus_frags),
           bac_shannon = quick_shannon(genus_frags)) %>% 
    #ungroup() %>% 
    select(sample_id, bac_chao1, bac_shannon) %>% 
    distinct() %>% 
    bind_rows(data.frame(sample_id = setdiff(include_samples, res_long$sample_id),
                         bac_chao1 = 0,
                         stringsAsFactors = F))
    
  
```


```{r}
p_bac_chao <- bac_alphadiv %>% 
  left_join(livestock_sample_metadata) %>% 
  #mutate(species = factor(species, levels = c("Chicken", "Pig", "Turkey", "Veal Calf", "Fish"))) %>% 
  #mutate(species = factor(species, levels = c("Broiler", "Pig", "Turkey", "Veal Calf", "Fish"))) %>% 
  ggplot(aes(x = explicit_country, y = bac_chao1, fill = species)) +
  geom_boxplot(outlier.shape = 21) +
  facet_grid(cols = vars(species), scales = "free_x", space = "free_x") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Country", y = "Bacterial richness (Chao1)") +
  guides(fill = F) +
  theme_classic(base_size = 9) +
  theme(panel.border = element_rect(linetype = "solid", fill = NA),
        strip.background = element_blank(),
        axis.line = element_blank())

p_bac_chao
```

```{r}
p_bac_evenness = bac_alphadiv %>% 
  filter(! is.na(bac_shannon)) %>% 
  left_join(livestock_sample_metadata) %>% 
  #mutate(species = factor(species, levels = c("Chicken", "Pig", "Turkey", "Veal Calf", "Fish"))) %>% 
  #mutate(species = factor(species, levels = c("Broiler", "Pig", "Turkey", "Veal Calf", "Fish"))) %>% 
  ggplot(aes(x = explicit_country, y = bac_shannon, fill = species)) +
  geom_boxplot(outlier.shape = 21) +
  facet_grid(cols = vars(species), scales = "free_x", space = "free_x") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Country", y = "Shannon index") +
  guides(fill = F) +
  theme_light(base_size = 9) +
  theme(panel.grid.major.x = element_blank())

#ggsave( "supp_shannon_bac_evenness.pdf",
#       width = 17.7, height = 6, units = "cm")

p_bac_evenness
```
## Compare bacterial and resistome diversity

```{r}
p_chao1_compare = livestock_sample_metadata %>% 
  left_join(bac_alphadiv) %>%
  left_join(res_alphadiv) %>%
  # Take out fish samples - too few to be meaningful
  filter(!species == "Fish") %>%
  ggplot(aes(bac_chao1, res_chao1, fill = species)) + 
  geom_point(shape = 21, size = 0.85) + 
  stat_smooth(span = 3, size = 0.5, col = "black") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Bacterial richness (Chao1)", y = "AMR richness (Chao1)") +
  guides(fill = F) +
  theme_classic(base_size = 9) +
  theme(panel.border = element_rect(linetype = "solid", fill = NA),
        strip.background = element_blank(),
        axis.line = element_blank())

p_chao1_compare
```

```{r}
p_evenness_compare = livestock_sample_metadata %>% 
  left_join(bac_alphadiv) %>%
  left_join(res_alphadiv) %>%
  filter(!species == "Fish") %>%
  na.omit() %>%
  ggplot(aes(exp(bac_shannon), exp(res_shannon), fill = species)) + 
  geom_point(shape = 21, size = 0.85) + 
  stat_smooth(span = 3, size = 0.5, col = "black") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Effective # bacterial species", y = "Effective # AMR genes") +
  guides(fill = F) +
  theme_classic(base_size = 9) +
  theme(panel.border = element_rect(linetype = "solid", fill = NA),
        strip.background = element_blank(),
        axis.line = element_blank())
  
p_evenness_compare
```

## Compare abundance and alpha diversity 
```{r}
# Compare AMR abundance and AMR richness
p_alr_even = res_long %>% 
  left_join(bact_hits) %>% 
  mutate(fully_normalised = frags_per_bp / bact_fragments) %>%
  group_by(sample_id) %>% 
  mutate(total_abundance = log2(sum(fully_normalised))) %>% 
  distinct(sample_id, total_abundance) %>% 
  left_join(livestock_sample_metadata) %>% 
  left_join(res_alphadiv) %>%
  filter(!species == "Fish") %>%
  ggplot(aes(exp(res_shannon), total_abundance, fill = species)) + 
  geom_point(shape = 21, size = 0.85) +
  stat_smooth(span = 3, size = 0.5, col = "black") +
  scale_fill_manual(values = plot_colours$species) +
  labs(x = "Effective # AMR genes", y = "Total AMR (log2 ALR)") +
  guides(fill = F) +
  theme_classic(base_size = 9) +
  theme(panel.border = element_rect(linetype = "solid", fill = NA),
        strip.background = element_blank(),
        axis.line = element_blank())

p_alr_even
```

# Make composite alpha diversity figure
```{r}
# pca_legend <- get_legend(p_pca)

alphadiv_multi_p <- plot_grid(p_chao, 
                           p_bac_chao, 
                           plot_grid(p_chao1_compare, p_evenness_compare, p_alr_even, 
                                     nrow = 1, labels = c("c", "d", "e")),
                           ncol = 1, rel_heights = c(0.5, 0.5, 0.7), labels = c("a", "b", NULL, NULL))

#ggsave("alphadiv_multi_p.pdf", plot = alphadiv_multi_p, width = 17.7, height = 17.7, units = "cm", device = cairo_pdf,
#       family = "Arial")

#ggsave("alphadiv_multi_p.png", plot = alphadiv_multi_p, width = 17.7, height = 17.7, units = "cm",
#       family = "Arial")

```


## Other data products

An 'FPKM' table for direct comparison to abundances based on the previous databases, alignment software etc.
```{r export_fpkm_product}
res_fpkm = res_long %>% 
  left_join(bact_hits) %>% 
  mutate(fpkm = frags_per_bp / bact_fragments * 10^9) %>%
  group_by(sample_id) %>%
  select(sample_id, gene_id, fpkm) %>%
  pivot_wider(names_from = gene_id, values_from = fpkm) %>%
  mutate_if(is.numeric, funs(replace_na(., 0)))

#write.table(res_fpkm, file = "output/resfinder.fpkm.wide.table.txt", sep="\t", quote = F)
```

```

